Better/correct handling of long button presses, DO NOT MERGE

olliw42 opened the pull request:
this resolves https://github.com/EdgeTX/edgetx/issues/52.

it actually turns out that 2.4 has the very same event handling as 2.3, and thus also has all the key handling it needs. All what was wrong is two bugs. These are corrected here.

olliw42 commented:
a killEvents() was missing, and I forgot that it should be break&long and not just break

can someone pl test on a Tx16S?

olliw42 commented on radio/src/gui/colorlcd/view_main.cpp:
```diff
@@ -257,12 +257,12 @@ bool ViewMain::onTouchEnd(coord_t x, coord_t y)
 {
   switch (event) {
-    case EVT_KEY_BREAK(KEY_MODEL):
+    case EVT_KEY_LONG(KEY_MODEL):
```
I find it MUUUUUCCCCCCCHHHHH more intuitive this way

with just break it will act only if you release the key, even if you press it for ages to happen ... it was kind of the first thing I noticed with etx and immediately concluded it is flawed, because that's what one gets from "cheap" key implementations ...

as said somewhere, if you wanted to have it short, then you should catch it for BOTH break and long

so, whatever the goal, is, as it is currently it is bad :)

btw, note, that the change wasn't then consistent, since for TELEM it still was long ...

olliw42 replied:
the cleanest for handling short presses would actually be to add an event short or click which fires before long...

raphaelcoeffic replied:
> btw, note, that the change wasn't then consistent, since for TELEM it still was long ...

true, but at the same time, it is not the kind of action you use very often, once setup. And it can be accessed from the main menu as well.

raphaelcoeffic replied:
I was hoping to use the short press on TELEM for something really related to telemetry.

olliw42 replied:
well, I can understand why you/some want to have it rather short than long ... it's however against my genes which ask for "symmetry", "consistency", "logic" or however one wants to call it to have one button working differently for no current reason ...

if you think most will prefer short, I'm happy to change to this, but I think you would then want to have all buttons behave that way (at least for as long as there is nothing happening on a short TELEM)

raphaelcoeffic replied:
Ok, so let's have it "short", and give the TELEM short press a real function, as the screen setup has nothing to do with "telemetry", right?

Btw. when implementing "short press", how do you want to allow short and long being two different functions if you don't wait for the release? That wouldn't work, would it?

raphaelcoeffic replied:
> can someone pl test on a Tx16S?

By the way, you should have a look at your discord PMs ;-)

olliw42 replied:
> Btw. when implementing "short press", how do you want to allow short and long being two different functions if you don't wait for the release? That wouldn't work, would it?

short would fire before long
now it's FIRST, LONG, REPEAT, REPEAT, ..., BREAK, with short it would be FIRST, SHORT, LONG, REPEAT, REPEAT, ..., BREAK, that's all

for long you are not waiting for break even now, the "release" is achieved by calling killEvents(), would be very much so also for short
it would be actually bloody easy (if I don't miss a crucial point) to add short ....

olliw42 replied:
probably one doesn't even have to add short, first might do it, I'll try ...

olliw42 replied:
yeah, first is doing it ... :)

olliw42 replied:
this was not what I was trying to imply ... I just meant that this is how it works with the long

raphaelcoeffic commented on radio/src/gui/colorlcd/tabsgroup.cpp:
```diff
@@ -169,12 +169,17 @@ void TabsGroup::onEvent(event_t event)
 {
   TRACE_WINDOWS("%s received event 0x%X", getWindowDebugString().c_str(), event);
 
+#if defined(KEYS_GPIO_REG_DGUP)
```
"s/DGUP/PGUP"

raphaelcoeffic commented on radio/src/gui/colorlcd/tabsgroup.cpp:
```diff
@@ -169,12 +169,17 @@ void TabsGroup::onEvent(event_t event)
   }
-#if defined(KEYS_GPIO_REG_UP)
-  else if (event == EVT_KEY_BREAK(KEY_PGUP)) {
+#if defined(KEYS_GPIO_REG_DGUP)
```
"s/DGUP/PGUP"

olliw42 commented:
this has it now all first
works all great on my T16
should better be tested on tx16s before merge

olliw42 commented:
@rotorman since you own a tx16s, would you think you could test this pr? it's just to check that the page up/page down buttons work for it

rotorman commented:
Can do. The buttons are labeled "page >" and "page <" not up/down, but will test them ASAP. Do you need just a test that they do anything (can navigate the menu back-and-forth), or do you need a test in a specific menu or a specific way of the button press (long/short)?

olliw42 commented:
yeah, the label might be different to their name in code

in the main screen, sys screen, mdl screen, telem screen, they should respond like the sys, mdl, telm, rtn buttons, i.e., a press should result in an action even if the button is held down for 10secs...

parts of the code are branched by defines which depend on t16 or tx16s, so my testing on t16 doesn't tell if it's all ok for tx16s, hence an explicit confirmation would be great

rotorman commented:
Built and flashed this PR to TX16S. I noticed that your change resulted in navigation with page> and page< executed immediately at key-down/button press (as in OTX 2.3) and not when button is released (previous ETX behavior). This indeed feels more nice & snappy. Also, both page< and page> buttons do execute the navigation, as expected on TX16S.

If the desire in EdgeTX is to use different functions for short and long key presses, then executing the action on first key down event might already execute an event that was not desired, if long press was intended.

For me, it is not clear yet what functions would be executed when long pressing the buttons and IMHO this issue might need more discussion on how to proceed according to the planned user interface concept for EdgeTX.

raphaelcoeffic commented:
> For me, it is not clear yet what functions would be executed when long pressing the buttons and IMHO this issue might need more discussion on how to proceed according to the planned user interface concept for EdgeTX.

Let's try to do this incrementally. For now, there is no clear idea what a long press could effect, and it does not seem to be necessary.

From my point if view, the only thing required for this would to fix the main menu: it uses BREAK rather than FIRST, so that the key event is caught first and triggers another main menu on top of the first one.

rotorman commented:
I made a short video about the main menu. Every roller press makes the main menu more dark and also getting out of the main menu required the same amount of return presses. I hope the video here shows this:

https://www.youtube.com/watch?v=IIauQJsydPg

pfeerick commented:
lol... I was about to post that I've just tested this on my TX16S, and it is working just fine... except for that infinite menu loop. Anyway, here is some video showing the keys otherwise responding as expected. It feels a lot smoother, especially as even if you are doing a short press and release, it responds on the press, rather than the release. ;) 

https://youtu.be/iMUkv6bYX5g

btw, since we're all on the discord group as well, you can use the roles to ping people with those transmitters... although the TX16 one loops T16 owners also unfortunately... `@TX16s/T16` ... still, at least you could ping people with that tag to test a PR like this one as needed (if they can)... might get more eyes on it for feedback in a more timely manner ;)

olliw42 commented:
> If the desire in EdgeTX is to use different functions for short and long key presses, then executing the action on first key down event might already execute an event that was not desired, if long press was intended.

this was why I initially had chosen LONG for them all, as this is how it is also on otx2.3 ... but @raphaelcoeffic suggested that it should be rather short since the long seemed to have confused users ... since there is no obvious action for long for any of them currently it was suggested to change it later if ever needed (summary of discussion of some above posts)

olliw42 commented:
concerning the main menu issue, reported by you all for tx16s

I am puzzled

he changes made here should not have had any effect on main_view, as much as I see
* ViewMainMenu directly derives from Window
* the changes made for TabsGroup are specific to tabsgroup, and TabsGroup is also derived directly from Window
* the changes made for ViewMain are specific to viewmain, and ViewMain is also derived directly from Window
* finally, ViewMainMenu has actually no handler of these buttons except exit itself, it must be somehow via SelectFabCarousel -> FormGroup -> ???

so how the heck do the changes made here affect the behavior of ViewMainMenu

I'm not saying that they don't, they obviously somehow do, but I'm somehow suspecting that the issue is actually due to another yet identified issue which this pr just exposes

I should add that in my observations the ViewMainMenu wasn't working very well anyhow with tabs, I could put it into states where it should not have been, like e.g. disappearing red select box etc ... could be related to what is found here

rotorman commented:
Here is the result of the old state d3189b38e6862f60075a1cd07bc11a1e946c592e (sorry for the 90° rotated video):
https://www.youtube.com/watch?v=KztVZTz5FQE
You can see that there I cannot even press multiple times the roller, but a correct selected action is already executed.
Also note the existence of the red selection rectangle, that seems to be missing presently.

olliw42 commented:
I come to the conclusion that there is something fundamentally flawed with libopengui and keys

I did four simple tests

1. I reverted the `case EVT_KEY_FIRST(KEY_ENTER):` back to `case EVT_KEY_BREAK(KEY_ENTER):` in view_main.cpp, but did so ONLY for this one key. The rational was that the libopenui form uses EVT_KEY_BREAK too, but does NOT do any eventKills (which for BREAK one should think should not be needed), and because this is what it was before for this key. This appears to work as it should.

2. I reverted the `case EVT_KEY_FIRST(KEY_ENTER):` back to `case EVT_KEY_BREAK(KEY_ENTER):` in view_main.cpp, and changed to  `case EVT_KEY_FIRST(KEY_ENTER):` in view_main_menu.h. The idea was to see how the focused view behaves. RTN doesn't work anymore at all with the main menu open, and it can hang. With the main menu open, pressing sys,mdl,telm opens the respective pages, i.e. the focused view do not appear to capture them.

3. I only changed to  `case EVT_KEY_FIRST(KEY_ENTER):` in view_main_menu.h, so that it is FIRST in both. Result similar to 2.

4. I went back to vanilla etx. You then also do strange things. E.g., open main menu, press sys, press rtn, press return ... and one can't get the menu to go away anymore.

this does NOT make any sense, since in principle the focused view should capture the events and underlying views should not see them (unless let through) ... so something with passing on and through events appears to be wrong

without understanding the issue, I conclude that in order to somehow operate with current libopenui one can not use FIRST, but always must use BREAK ... arrrrgggghhhh

so, this PR is nonsense for current libopenui. 
To close or not to close?

olliw42 commented:
so, yet another (last) test

I changed all EVT_KEY_XXXX(xxx) to ONLY EVT_KEY_BREAK(xxx) in view_main.cpp and tabsgroup.cpp. The idea was that if it indeed would be true that libopenui only uses BREAK also only using BREAK in the higher levels might avoid some out-of-sync events handling ... and it's still all incorrect.

With that change I did this key strokes

- press enter => main menu opens
- scroll => red box moves
- press sys => tools page opens, note: the tools page is often screwed up like in #60 but "much worse" in the sense that it's not just a 9 showing
- press rtn => tools page disappears and main menu is back, BUT wo red selection box, AND it does not react to anything anymore, dead
- press enter again =>another main menu opens, it is very clearly another main menu, see picture


![IMG_20210523_074802](https://user-images.githubusercontent.com/6089567/119249924-eb2fc200-bb9c-11eb-9273-27df62451dfb.jpg)



so, I think it hardly can be clearer now that for libopenui or the higher levels using it it holds that either or both
- the event handling it is fundamentally flawed, in that it is not restricted properly
- the drawing is fundamentally flawed, in that it doesn't clear properly

I would not be surprised to learn that all the flickering  and temporarily miss-drawings we all have observed is related to that

to me it seems that one needs careful revisiting of libopenui, or if it is very strongly believed that libopenui is ok, of understanding how it is supposed to be properly used and the higher levels being adjusted accordingly.

I'm not sure one should close this PR, it might be useful to have it around as resource. I hence have marked it as DO NOT MERGE.

raphaelcoeffic commented:
Code has been merged in #68
